### PR TITLE
Final refactor

### PR DESCRIPTION
**Pull request by ScottRhodie**
Code has been refactored from the original code and now into something clearer and more readable.

### CODE REVIEW

File: app/routes.js [hunk 1]
```diff
@@ -1,43 +1,119 @@
...
+                        });
+                    }
+                })
         })
 
 
-        app.get('/api/v1/users/', (req, res) => {
-            return controller.findAllUsers(req)
+        app.put('/api/v1/users/:userId', (req, res) => {
+            return controller.findUserByIdAndUpdate(req)
```
Comment (feliun) [inline app/routes.js]: `updateUser` is better naming

File: app/routes.js [hunk 2]
```diff
@@ -1,43 +1,119 @@
...
+                .catch(err => {
                     res.status(err.errorCode || 500).send({
                         message: err.message || "Some error occurred while creating the user."
                     });
                 });
+        });
+
+
+        app.get('/api/v1/users/:userId', (req, res) => {
+            return controller.findSingleUser(req)
```
Comment (feliun) [inline app/routes.js]: `findUser` is better naming

File: app/routes.js [hunk 3]
```diff
@@ -1,43 +1,119 @@
...
+                        return res.status(404).send({
+                            error: "There is no user with the ID of : " + req.params.userId
+                        });
+                    }
+                })
+        })
+
+
+        app.delete('/api/v1/users/:userId', (req, res) => {
+            return controller.findUserByIdAndDelete(req)
```
Comment (feliun) [inline app/routes.js]: `deleteUserById`

File: app/controller.js [hunk 4]
```diff
@@ -4,12 +4,31 @@ module.exports = store => {
     return store.saveUser(user);
   };
 
+
   const findAllUsers = Users => {
     return store.findAllUsers(Users)
   }
 
+
```
Comment (servatj) [inline app/controller.js:12]: I think in this case you don't have to wrap an existing method in another function with the same name. I think in this case you can directly use the store wherever you need injecting them as you did in this module. Are you trying to use the proxy pattern here for some reason?  The proxy pattern can make sense if you want to protect the store by validating or transforming things in the functions you're wrapping...

Comment (feliun) [reply app/controller.js:12]: @servatj you're totally correct. However this was done so @ScottRhodie can get his head around the concept of a controller as proxy that could do more than that

File: app/routes.js [hunk 5]
```diff
@@ -1,43 +1,119 @@
...
+                "Welcome to Instagram! <br> Check /api/v1/users/[userId] to view users by ID."
+            );
         });
 
-        app.post('/api/v1/users/', (req, res) => {
-            return controller.createUser(req.body)
+
+        app.get("/api/v1/users/", (req, res) => {
+            return controller
+                .findAllUsers(req)
```
Comment (servatj) [inline app/routes.js]: here you are using your controller functions that wraps the store, it probably make more sense to create a method getAllUser in the controller that uses getFindUsers from the store. Anyway i don't like the name of this function sin when you want to ALL you're actually not Finding o Searching anything

File: app/controller.js [hunk 6]
```diff
@@ -4,12 +4,31 @@ module.exports = store => {
...
+  const findSingleUser = userId => {
+    return store.findSingleUser(userId)
+  }
+
+
+  const findUserByIdAndUpdate = userId => {
```
Comment (servatj) [inline app/controller.js]: You can simply this : 
```
const findUserByIdAndDelete = userId => store.findUserByIdAndDelete(userId)
```
  
```
  const findUserByIdAndDelete = store.findUserByIdAndDelete
```

Probably it's not super important that you know what point free is at the moment , but I think it's going to help you a lot when readers other functional code --> https://lucasmreis.github.io/blog/pointfree-javascript/

File: app/routes.js [hunk 7]
```diff
@@ -1,43 +1,119 @@
...
+                    res.status(500).send({
+                        message: err.message || "Unable to rerieve users"
+                    });
+                });
+        });
+
+
+        app.post("/api/v1/users/", (req, res) => {
+            return controller
+                .createUser(req.body)
```
Comment (servatj) [inline app/routes.js:35]: again about readability probably is better 
 
   let newUserId = reuquest.body.userId 
  ...
  .createUser(newUserId)

File: app/routes.js (hunk 2 above)
Comment (servatj) [inline app/routes.js]: same here I think you get the point 
 const userId =  req.body.userId 
....

   findSingleUser(userId)  much more readable

File: app/routes.js (hunk 3 above)
Comment (servatj) [inline app/routes.js]: Good naming here but be careful with functions that do more than one thing, in this case it's super clear that the function first looks for a user and then it deletes it...  the think is that most databases delete based on query directly.,, so probably you can name this functions as deleteById(userId)

File: app/store.js [hunk 8]
```diff
@@ -18,16 +18,35 @@ module.exports = ({
...
+                    .then((response) => {
                         console.log("Listing all users now...");
-                        return responce;
+                        return response;
                     })
-            }
+            },
 
 
+            findSingleUser: (req) => {
```
Comment (servatj) [inline app/store.js]: again not necessary to wrap but if you want to wrap it as you did put userId in the params instead of request...

**Review by servatj — APPROVED** [review]
Looks quite good and we can see a lot of progress there. I posted some comments though, I think they can be in you interest, don't hesitate to comment or ask me any question :)

File: app/routes.js (hunk 1 above)
Comment (ScottRhodie) [reply app/routes.js]: Change implemented.

File: app/routes.js (hunk 3 above)
Comment (ScottRhodie) [reply app/routes.js]: Change implemented.

File: app/routes.js (hunk 2 above)
Comment (ScottRhodie) [reply app/routes.js]: Change implemented. Thanks for the suggestion.  Going to try keep the naming more simple and to the point.

File: app/controller.js (hunk 4 above)
Comment (servatj) [reply app/controller.js:12]: Sure I see !!! As iImentioned I think he progressed a lot :)

File: app/routes.js (hunk 5 above)
Comment (ScottRhodie) [reply app/routes.js]:
```
app.get("/api/v1/users/", (req, res) => {
            return controller
                .getAllUsers(req)
```
I've changed the naming on this to getAllUsers

File: app/controller.js (hunk 6 above)
Comment (ScottRhodie) [reply app/controller.js]: Discussed this over slack and we're going to hold off on this and maybe come back to it in the future as I'm just starting to get the hang of this now and about to jump into building the frontend in react, and I'm worried about overloading with information.  Pinned this however in the convo and will be coming onto it for sure.

NDR : Be worth checking out https://tech.io/playgrounds/2980/practical-introduction-to-functional-programming-with-js/pure-functions also as recommended

File: app/routes.js (hunk 3 above)
Comment (ScottRhodie) [reply app/routes.js]: This has been renamed to deleteUserById

File: app/store.js (hunk 8 above)
Comment (ScottRhodie) [reply app/store.js]: Applied this to updateUser, findUser & deleteUserById.  Is that okay?

**Comment by ScottRhodie** [comment]
Thank you a lot for the help and advice @feliun & @servatj 

Merging pull request.

**Pull request merged**